File: longestSubstringWithKDistinctCharacters.py
def longestSubstringWithKDistinctCharacters(str, k):
    i,j,max_count=0,0,0
    ump = {}
    while j<len(str):
        if len(ump)<=k:
            if str[j] not in ump:
                ump[str[j]]=1
            elif str[j] in ump:
                ump[str[j]]+=1
            if(len(ump)==k):
                max_count=max(max_count,j-i+1)
            j+=1
        else:
            ump[str[i]]-=1
            if ump[str[i]]==0:
                del ump[str[i]]
            i+=1
    return -1 if max_count==0 else max_count
    



print(longestSubstringWithKDistinctCharacters('aabcdefff', 3))
# 5 (because 'defff' has length 5 with 3 characters)

File: longestSubstringWithKDistinctCharacters.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

int longestSubstringWithKDistinctCharacters(string str, int k)
{
    int i = 0, j = 0;
    int max_count = 0;
    unordered_map<char, int> ump;
    while (j < str.size())
    {
        if (ump.size() <= k)
        {
            ump[str[j]]++;
            if (ump.size() == k)
            {
                max_count = max(max_count, j - i + 1);
            }
            j++;
        }
        else
        {
            ump[str[i]]--;
            if (ump[str[i]] == 0)
                ump.erase(str[i]);
            i++;
        }
    }
    return max_count > 0 ? max_count : -1;
}

int main()
{
    int cases;
    cin >> cases;
    while (cases--)
    {
        string str;
        int k;
        cin >> str >> k;
        cout << longestSubstringWithKDistinctCharacters(str, k) << endl;
    }
}

// given "abcbbbbcccbdddadacb", the longest substring that contains 2 unique character is "bcbbbbcccb".